File: Project_Nr_1/Part_1/Receiver/ReceiverSocket.cs
using System.Text;
using System.Net.Sockets;
using System.Net;
using Message_Agent.Common;
using System.Collections.Generic;

namespace Receiver
{
    class ReceiverSocket
    {
        private Socket _socket;
        private List<string> _topics;

        public ReceiverSocket()
        {
            _topics = new List<string>();

            _socket = new Socket(
                AddressFamily.InterNetwork,
                SocketType.Stream,
                ProtocolType.Tcp
            );
        }

        public void Connect(string ipAddress, int port)
        {
            _socket.Connect(new IPEndPoint(IPAddress.Parse(ipAddress), port));
            Console.WriteLine("Receiver connected to broker");
        }

        
        public bool Authenticate(string username, string password, bool isSignUp)
        {
            string command = isSignUp ? "signup" : "signin";
            var data = Encoding.UTF8.GetBytes($"{command}#{username}#{password}");
            _socket.Send(data);

            byte[] buffer = new byte[1024];
            int received = _socket.Receive(buffer);
            string response = Encoding.UTF8.GetString(buffer, 0, received);

            if (response.StartsWith("authOK"))
            {
                var parts = response.Split('#');

                if (parts.Length > 1 && !string.IsNullOrEmpty(parts[1]))
                {
                    foreach (var topic in parts[1].Split(','))
                    {
                        if (!_topics.Contains(topic))
                            _topics.Add(topic);
                    }
                }

                StartReceive();
                return true;
            }

            Console.WriteLine($"Authentication failed: {response}");
            return false;
        }

        public void Subscribe(string topic)
        {
            if (_topics.Contains(topic))
            {
                Console.WriteLine($"Already subscribed to: {topic}");
                return;
            }

            _topics.Add(topic);

            var data = Encoding.UTF8.GetBytes("subscribe#" + topic);
            Send(data);

            Console.WriteLine($"Subscribed to: {topic}");
        }

        public List<string> GetTopics()
        {
            return _topics;
        }

        private void StartReceive()
        {
            ConnectionInfo connection = new ConnectionInfo();
            connection.Socket = _socket;

            _socket.BeginReceive(
                connection.Data,
                0,
                connection.Data.Length,
                SocketFlags.None,
                ReceiveCallBack,
                connection
            );
        }

        private void ReceiveCallBack(IAsyncResult asyncResult)
        {
            ConnectionInfo connectionInfo =
                asyncResult.AsyncState as ConnectionInfo;

            try
            {
                SocketError response;

                int buffSize =
                    _socket.EndReceive(asyncResult, out response);

                if (response == SocketError.Success && buffSize > 0)
                {
                    byte[] payloadBytes = new byte[buffSize];

                    Array.Copy(
                        connectionInfo.Data,
                        payloadBytes,
                        payloadBytes.Length
                    );

                    PayloadHandler.Handle(payloadBytes);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(
                    $"Can't receive data from broker. {e.Message}"
                );
            }
            finally
            {
                try
                {
                    connectionInfo.Socket.BeginReceive(
                        connectionInfo.Data,
                        0,
                        connectionInfo.Data.Length,
                        SocketFlags.None,
                        ReceiveCallBack,
                        connectionInfo
                    );
                }
                catch (Exception e)
                {
                    Console.WriteLine($"{e.Message}");
                    connectionInfo.Socket.Close();
                }
            }
        }

        private void Send(byte[] data)
        {
            try
            {
                _socket.Send(data);
            }
            catch (Exception e)
            {
                Console.WriteLine(
                    $"Could not send data: {e.Message}"
                );
            }
        }
    }
}

File: Project_Nr_1/Part_1/Python_Receiver/receiver_socket.py
import socket
import threading

from payload_handler import PayloadHandler

class ReceiverSocket:
    def __init__(self):
        self._socket = socket.socket(
            socket.AF_INET,
            socket.SOCK_STREAM
        )

        self._topics = []

    def connect(self, ip_address, port):
        self._socket.connect((ip_address, port))
        print("Receiever connected to broker")

    def authenticate(self, username, password, is_sign_up):
        command = "signup" if is_sign_up else "signin"

        data = f"{command}#{username}#{password}".encode("utf-8")

        self._socket.send(data)

        response = self._socket.recv(1024).decode("utf-8")

        parts = response.split("#")

        if parts[0] == "authOK":
            if len(parts) > 1 and parts[1]:
                for topic in parts[1].split(","):
                    if topic not in self._topics:
                        self._topics.append(topic)

            self.start_receive()

            return True, "Authenticated succcessfully."

        reason = parts[1] if len(parts) > 1 else "unknown error"

        return False, reason

    def subscribe(self, topic):
        if topic in self._topics:
            print(f"Already subscribed to: {topic}")
            return

        self._topics.append(topic)

        data = f"subscribe#{topic}".encode("utf-8")

        self.send(data)

        print(f"Subscribed to: {topic}")

    def get_topics(self):
        return self._topics

    def start_receive(self):
        receive_thread = threading.Thread(
            target = self.receive_messages,
            daemon=True
        )

        receive_thread.start()

    def receive_messages(self):
        while True:
            try:
                payload_bytes = self._socket.recv(4096)

                if not payload_bytes:
                    break

                PayloadHandler.handle(payload_bytes)

            except Exception as e:
                print(f"Can't receive data from broker. {e}")
                break

    def send(self, data):
        try:
            self._socket.send(data)
        except Exception as e:
            print(f"Could not send data: {e}")